parachain/pallets/omni-account/src/lib.rs

```diff
@@ -0,0 +1,218 @@
+		type MaxIDGraphLength: Get<u32>;
+	}
+	pub type IDGraphMembers<T> = BoundedVec<MemberAccount, <T as Config>::MaxIDGraphLength>;
+
+	#[pallet::storage]
+	pub type LinkedIdentityHashes<T: Config> =
+		StorageMap<Hasher = Blake2_128Concat, Key = H256, Value = ()>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn id_graphs)]
+	pub type IDGraphs<T: Config> =
+		StorageMap<Hasher = Blake2_128Concat, Key = Identity, Value = IDGraphMembers<T>>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn id_graph_hashes)]
+	pub type IDGraphHashes<T: Config> =
+		StorageMap<Hasher = Blake2_128Concat, Key = Identity, Value = H256>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Identity linked
+		IdentityLinked { who: Identity, identity: H256 },
+		/// Identity remove
+		IdentityRemoved { who: Identity, identity_hashes: Vec<H256> },
+		/// Identity made public
+		IdentityMadePublic { who: Identity, identity_hash: H256 },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Identity is already linked
+		IdentityAlreadyLinked,
+		/// IDGraph len limit reached
+		IDGraphLenLimitReached,
+		/// Identity not found
+		IdentityNotFound,
+		/// Invalid identity
+		PrimeIdentityInvalid,
+		/// Prime identity not found
+		PrimeIdentityNotFound,
+		/// Identity is private
+		IdentityIsPrivate,
+		/// Identities empty
+		IdentitiesEmpty,
+		/// IDGraph hash does not match
+		IDGraphHashMismatch,
+		/// Missing IDGraph hash
+		IDGraphHashMissing,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight((195_000_000, DispatchClass::Normal))]
+		pub fn link_identity(
+			origin: OriginFor<T>,
+			who: Identity,
+			identity_hash: H256,
+			identity: MemberIdentity,
+			maybe_id_graph_hash: Option<H256>,
+		) -> DispatchResult {
+			let _ = T::TEECallOrigin::ensure_origin(origin)?;
+			ensure!(
+				!LinkedIdentityHashes::<T>::contains_key(identity_hash),
+				Error::<T>::IdentityAlreadyLinked
+			);
+			let mut id_graph_members = match IDGraphs::<T>::get(&who) {
+				Some(id_graph_members) => {
+					let current_id_graph_hash =
+						IDGraphHashes::<T>::get(&who).ok_or(Error::<T>::PrimeIdentityNotFound)?;
+					if let Some(id_graph_hash) = maybe_id_graph_hash {
+						ensure!(
+							current_id_graph_hash == id_graph_hash,
+							Error::<T>::IDGraphHashMismatch
+						);
+					} else {
+						return Err(Error::<T>::IDGraphHashMissing.into());
+					}
+					id_graph_members
+				},
+				None => {
+					let who_hash = who.hash().map_err(|_| Error::<T>::PrimeIdentityInvalid)?;
+					if LinkedIdentityHashes::<T>::contains_key(who_hash) {
+						return Err(Error::<T>::IdentityAlreadyLinked.into());
+					}
+					let mut id_graph_members: IDGraphMembers<T> = BoundedVec::new();
+					id_graph_members
+						.try_push(MemberAccount {
+							id: MemberIdentity::from(who.clone()),
+							hash: who_hash,
+						})
+						.map_err(|_| Error::<T>::IDGraphLenLimitReached)?;
```

Comment:
Don't we need to push both PrimeIdentity and MemberAccount?

Reply:
In this part of the code we are only creating the IDGraph (for the first time), this is why we only add the prime identity. In the code below is where we add identity (MemberAccount). This is why the `id_graph_members` is initialized as `mut`, this way we don't have to repeat the logic depending on the branch, we just add the new member one time for both cases

Reply:
I've refactored the code to make it clear